Generate certificate guide (#12847)
microsoftgraph/microsoft-graph-docs: concepts/generatecertificate.md

---

> title: "Generate a self-signed public certificate for your application"

Reviewer: ```suggestion
title: "Generate a self-signed public certificate for authentication"
```

> description: "Generate a self-signed public certificate to authenticate your application."

Reviewer: ```suggestion
description: "Create a self-signed public certificate to authenticate your application."
```

author: "FaithOmbongi"
localization_priority: Priority
---

> # Generate a self-signed public certificate to authenticate your application

Reviewer: ```suggestion
# Generate a self-signed public certificate for authentication
```

> Azure Active Directory (Azure AD) supports two types of authentication for service principals: **password-based authentication** (app secret) and **certificate-based authentication**. While app secrets can easily be generated via the **App registrations** blade, Microsoft recommends using a certificate.

Reviewer: ```suggestion
Azure Active Directory (Azure AD) supports two types of authentication for service principals: **password-based authentication** (app secret) and **certificate-based authentication**. While app secrets can easily be generated by using the **App registrations** blade in the portal, it is recommended that your application uses a certificate.
```

> For testing purposes, you can use a self-signed public certificate instead of a Certificate Authority (CA) signed certificate. This article shows you how to use PowerShell to generate and export a self-signed certificate with or without a private key.

Reviewer: ```suggestion
For testing purposes, you can use a self-signed public certificate instead of a Certificate Authority (CA) signed certificate. This article shows you how to use PowerShell to create and export a self-signed certificate with or without a private key.
```

> [!CAUTION]
> Using a self-signed certificate is only recommended for development, not production.

> Modern versions of Windows (Windows 8.1 and greater, and Windows Server 2012R2 and greater) include a built-in PowerShell cmdlet `New-SelfSignedCertificate` to create a self-signed certificate and the `Export-Certificate` cmdlet to export it to a location that is easily accessible. This tutorial uses these built-in PowerShell tools to generate and export the certificate.

Reviewer: ```suggestion
Modern versions of Windows (Windows 8.1 and greater, and Windows Server 2012R2 and greater) include a built-in PowerShell cmdlet `New-SelfSignedCertificate` to create a self-signed certificate and the `Export-Certificate` cmdlet to export it to a location that is easily accessible. This page shows you options for using PowerShell to create the certificate that you need.
```

Reviewer 2: Or perhaps 
```
...create the certificate and exports of the portions of the certificate that you will need.
```

> > [!NOTE]

Reviewer: This doesn't need to be in Note format.

> + Longer **-KeyLength** values are supported, but 2048 bit size is highly recommended for the best combination of security and performance.
> + Only `RSA` is allowed for **-KeyAlgorithm**.
> + Supported values for **-HashAlgorithm** are `SHA256`, `SHA384`, and `SHA512`.

> > + This tutorial generates certificates that are valid for 1 year. To customize the start and expiry date as well as other properties, see the [`New-SelfSignedCertificate` reference](/powershell/module/pki/new-selfsignedcertificate?view=windowsserver2019-ps).

Reviewer: This isn't really a tutorial, I would just refer to it as a page. Or you could rewrite something like this: The certificates that the cmdlets generate are valid for 1 year. To customize the start and expiration date as well as other properties, see [New-SelfSignedCertificate](/powershell/module/pki/new-selfsignedcertificate?view=windowsserver2019-ps).

> > + Self-signed certificate generated following this tutorial are supported for use for both client and server authentication.

Reviewer: ```suggestion
> + The self-signed certificates that you generate are supported for use for both client and server authentication.
```

> ## Option 1: Generate and export your public certificate without a private key

Reviewer: ```suggestion
## Option 1:  Create and export your public certificate without a private key
```

> ## Option 1: Generate and export your public certificate without a private key

Reviewer 2: I traditionally think of headings with names like "option 1 and option 2" as showing two ways to achieve the *same* outcome.  But these section are achieving different outcomes  option 2 does more. E.g., if you are trying to use Azure Automation then you can't use option 1. I suggest that there should be some text that says which one you need to do. Something like:
```
This option will generate a key pair and store it locally, but only export the public portion of the certificate.
Use this option if you will only be authenticating to Azure AD from a service principal running on this PC.  For example, if you are just testing or experimenting with using Microsoft Graph PowerShell from your computer, then you do not need to export the private key.
However, if your service will be authenticating from other computers or from Azure Automation, then use Option 2.
```

Author: Sounds good.

> In an elevated PowerShell prompt, run the following command and leave the PowerShell console session open. In the following command, replace `{certificateName}` with name you wish to give your certificate.

Reviewer: ```suggestion
In an elevated PowerShell prompt, run the following command and leave the PowerShell console session open. In the following command, replace `{certificateName}` with the name that you wish to give to your certificate.
```

```powershell

$cert = New-SelfSignedCertificate -Subject "CN={certificateName}" -CertStoreLocation "Cert:\CurrentUser\My" -KeyExportPolicy Exportable -KeySpec Signature -KeyLength 2048 -KeyAlgorithm RSA -HashAlgorithm SHA256    ## Replace {certificateName}

```

The **$cert** variable in the previous command stores your certificate in the current session and allows you to export it. The command below exports the certificate in `.cer` format. You can export it in any file format supported on the Azure Portal including `.pem` and `.crt`.

```powershell

Export-Certificate -Cert $cert -FilePath "C:\Users\admin\Desktop\{certificateName}.cer"   ## Specify your preferred location and replace {certificateName}

```

Your certificate is now ready to upload to the Azure Portal. Once uploaded, you can retrieve the certificate thumbprint for use to authenticate your application.

> ## Option 2: Generate and export your public certificate with it's private key

Reviewer: ```suggestion
## Option 2: Create and export your public certificate with it's private key
```

> ## Option 2: Generate and export your public certificate with it's private key

Reviewer 2: I think it should be "and its private key" rather than "with it's private key".

> 

Reviewer 2: Might want to add a sentence about why youd choose this option, and clarify the differences between the two files, something like

```
If your service will be running on another cloud, or you will be using Azure Automation, you will need to export both the private key and the public key portions that you generated.  The private key will be written to a file with the PFX extension, and the public key will be written to a file with the CER extension.

The private key is needed by the component that is initiating the authentication.  For example, if using Azure Automation, you would upload the private key PFX file to Azure Automation in your subscription.  

The public key is needed by the component that is confirming the authentication.  For example, if authenticating as a Azure AD application, then you would upload the public key CER file to the Azure AD application registration.
```

Author: Good point. Added to the doc. But I wonder if we might need a graphic to illustrate this?

```

The **$cert** variable in the previous command stores your certificate in the current session and allows you to export it. The command below exports the certificate in `.cer` format. You can export it in any format supported on the Azure Portal including `.pem` and `.crt`.

```powershell

Export-Certificate -Cert $cert -FilePath "C:\Users\admin\Desktop\{certificateName}.cer"   ## Specify your preferred location and replace {certificateName}

```

Still in the same session, create a password for your certificate private key and save it in a variable. In the following command, replace `{myPassword}` with the password you wish to use to protect your certificate private key.

```powershell

$mypwd = ConvertTo-SecureString -String "{myPassword}" -Force -AsPlainText  ## Replace {myPassword}

```

Now, using the password you stored in the `$mypwd` variable, secure and export your private key.

```powershell

Export-PfxCertificate -Cert $cert -FilePath "C:\Users\admin\Desktop\{privateKeyName}.pfx" -Password $mypwd   ## Specify your preferred location and replace {privateKeyName}

```

> Your certificate (`.cer` file) is now ready to upload to the Azure Portal. You also have a private key (PFX file). Once uploaded, you can retrieve the certificate thumbprint for use to authenticate your application. 

Reviewer 2: Please add something about the handling of the PFX file.  The CER file can be read by others, but should not be writable.  The PFX file, while encrypted, should not be readable or writable by others.

Author: Noted. Added.

## Optional task: Delete the certificate from the keystore.

> If the application that will interact with Microsoft Graph and Azure AD will not run from your local machine, you can delete the certificate (and the private key) you generated from your personal store. First, get it's thumbprint.

Reviewer: ```suggestion
If the application that interacts with Microsoft Graph and Azure AD won't be run from your local machine, you can delete the certificate (and the private key) you generated from your personal store. First, get it's thumbprint.
```

```powershell

Get-ChildItem -Path "Cert:\CurrentUser\My" | Where-Object {$_.Subject -Match "{certificateName}"} | Select-Object Thumbprint, FriendlyName    ## Replace {privateKeyName} with the name you gave your certificate

```

> Then, copy the thumbprint that is displayed and use it delete the certificate and it's private key.

Reviewer: ```suggestion
Then, copy the thumbprint that is displayed and use it to delete the certificate and it's private key.
```

```powershell

Remove-Item -Path Cert:\CurrentUser\My\{pasteTheCertificateThumbprintHere} -DeleteKey

```

## Next steps

> Now that you have your certificate in one of the formats supported by Azure AD, use it to [authenticate your application]().

Reviewer: Still waiting for this page?

Author: Yeah.

> 

Reviewer 2: Could you add a section on reminding them that this certificate will expire.  Something like:

```
## Plan to keep the certificate up to date

If you created a self-signed certificate following the steps above for authentication, keep in mind that the certificate will have a limited lifetime before it will expire. You will need to regenerate the certificate and upload the new certificate before its expiration date.

There are two places where you can see the expiration date in the Azure portal.
* If you are using Azure Automation, the **Certificates** screen on the Automation account displays the expiration date of the certificate.
* In Azure AD, on the app registration, the **Certificates & secrets** screen displays the expiration date of the certificate used for the Azure Automation account.

```

Author: Sounds good. Though changing the title because we don't give guidance for renewing the certificate.

## See more

> 

Reviewer: I think the link to the cmdlet reference is enough. The list isn't needed.

The `New-SelfSignedCertificate` cmdlet supports additional properties that you can use to configure your certificate. These include the following:
+ `-DnsName` to specify the issuer. This can be your domain address. 
+ `-NotBefore` and `-NotAfter` properties to explicitly specify when the certificate becomes valid and when it expires.
+ `-Pin` to specify the personal identification number (PIN) used to access the private key of the new certificate.

> To know the properties you can use to configure your certificate, see the [`New-SelfSignedCertificate` reference](/powershell/module/pki/new-selfsignedcertificate?view=windowsserver2019-ps).

Reviewer: ```suggestion
For more information about the properties that you can use to configure your certificate, see [New-SelfSignedCertificate](/powershell/module/pki/new-selfsignedcertificate?view=windowsserver2019-ps).
```

Author: Removed this list and added the link in the first section instead.
